src/extra/components/detector_motors.py

```diff
@@ -0,0 +1,414 @@
+            return xarray.DataArray(
+                pos, dims=dims, coords=coords, name=self._position_key)
+        else:
+            return pos
+
+    def _read_positions(self):
+        """Reads and compresses motor positions."""
+        if hasattr(self, "_positions"):
+            return self._positions
+
+        # read positions
+        pos = np.zeros((self.num_trains, self.num_sources), dtype=float)
+        for source_no, key_data in enumerate(self.keys):
+            pos[:, source_no] = key_data.ndarray()
+        pos = pos.reshape(-1, *self.shape)
+
+        # compress
+        axes = tuple(range(1, pos.ndim))
+        ix = np.flatnonzero(
+            np.insert(np.any(np.diff(pos, axis=0), axis=axes), 0, True))
+
+        # train ids when positions are changed
+        train_ids = np.asarray(self.dc.train_ids)[ix]
+        # count of trains at every position
+        counts = np.diff(np.append(train_ids.astype(int),
+                                   int(self.dc.train_ids[-1]) + 1))
+
+        self._positions = train_ids, pos[ix], counts
+        return self._positions
+
+    def compressed_positions(self, labelled=True):
+        """Returns the unique motor positions and corresponding train IDs.
+
+        Args:
+            labelled (bool):
+                If True, returns the xarray with labelled dimensions,
+                overwise returns tuple of numpy.ndarrays
+
+        Returns:
+            positions (tuple of two numpy.ndarray or xarray.DataArray):
+                The motor positions
+        """
+        train_ids, values, _ = self._read_positions()
+        if labelled:
+            dims = ["trainId"] + list(self.dims.keys())
+            coords = {"trainId": train_ids}
+            coords.update({name: values for name, values in self.dims.items()})
+            return xarray.DataArray(
+                values, dims=dims, coords=coords, name=self._position_key)
+        else:
+            return train_ids, values
+
+    def positions_at(self, tid):
+        """Returns motor positions at given train.
+
+        Args:
+            tid (int):
+                Train ID
+
+        Returns:
+            postions (numpy.ndarray):
+                The motor positions
+        """
+        train_ids, values, _ = self._read_positions()
+        i = np.searchsorted(train_ids, tid, side="right")
+        return values[max(i - 1, 0)]
+
+    @property
+    def most_frequent_positions(self):
+        """Returns most frequent motor positions."""
+        _, values, counts = self._read_positions()
+        return values[np.argmax(counts)]
+
+    @property
+    def first(self):
+        """Returns the first motor positions."""
+        tid = self.train_ids[0]
+        return self.positions_at(tid)
+
+    @property
+    def last(self):
+        """Returns the last motor positions."""
+        tid = self.train_ids[-1]
+        return self.positions_at(tid)
```

Comment:
I would make these regular methods, rather than properties, because calling them may need to load data. This is a judgment call, but generally properties (which you access like `.first` without an explicit call) are expected to be things that are cheap to compute.

If I see methods like `.first()`, or even `.positions_at()`, I'd also assume that these load only the required bit of data, so they're faster than loading all the data and indexing it. But these work by loading all the data, which seems a bit surprising.

Reply:
I would really surprised if the loading of one number is really much faster than load entire vector. Of course there could be a corner cases.

I will remove first and last at all. I don't really think that they are useful in the interface. For the `position_at`, I think it is reasonable compromise to read a vector once.